### DISCUSSION TITLE
Equipment and heard rate zone

### DISCUSSION THREAD

**Discussion by Seb3925** [tcgoetz/GarminDB#142] [Q&A] [answered]
Hello,
Thanks for the DB, it's really a good job!!  Is it possible to also download for each activity the equipment used (shoes, bike...) and the duration in each heard rate zone (in time format)?

Thanks in advance for your feedback

Seb

**Comment by smros** [comment]
Hello, I'm also interested in a feature like this.  I'd love to calculate the time I spend in each HR zone and also be able to match it against the activities at the time.  Most of my activities are training/cardio-training and I'd like to track over time.

The garmin Connect App can generate it, but only for views in the app.  I'd to pull out the full historical data myself.   

I assume the steps are:
For each activity:

1. Calculate my HR ranges at the that time
2. Get the times from the DB ( I can see in the Activities Table a start and end time)
3. Get the HR measurments from those times (I presume this is available)
4. Calc the required stats
5. Store in a new table

I think this can be done with what we have.  I have some Python, SQL and JupyterLab experience but to attack this problem here, what's the best place to start?  I saw in some other discussions that instead of querying directly, there are "classes" to use to access the data.  Is that the place to start?

-Steve

**Reply by tcgoetz** [reply]
> Hello, I'm also interested in a feature like this. I'd love to calculate the time I spend in each HR zone and also be able to match it against the activities at the time. Most of my activities are training/cardio-training and I'd like to track over time.
> 

The data is already present in the FIT file. I have started adding support for it. Here's an example of th time in zones data from one of my FIT files:

```
Message: DataMessage(<MessageType.hr_zones_timer: 216>: [timestamp(datetime.datetime(2021, 11, 25, 12, 41, 21, tzinfo=datetime.timezone.utc) (1006778481)), time_per_zone([datetime.time(0, 1, 9, 910000), datetime.time(0, 1, 2, 1000), datetime.time(0, 0, 39, 999000), datetime.time(0, 3, 27, 998000), datetime.time(0, 3, 49, 3000), datetime.time(0, 0), datetime.time(0, 0)] ([69910, 62001, 39999, 207998, 229003, 0, 0])), unknown_5([invalid] ([4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295])), unknown_9([invalid] ([65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535])), hr_zones_timer_type(<HeartRateZonesTimerType.lap: 19> (19)), record_num(0), unknown_15([invalid] (65535)), hr_zones([124, 136, 148, 162, 175, 188]), hr_zones_method(<HeartRateZonesMethod.heart_rate_reserve: 2> (2)), max_heart_rate(188.0 bpm), min_heart_rate(58.0 bpm), lactate_threshhold_heart_rate(162.0 bpm), unknown_14([invalid] (255))])
Message: DataMessage(<MessageType.hr_zones_timer: 216>: [timestamp(datetime.datetime(2021, 11, 25, 12, 50, 27, tzinfo=datetime.timezone.utc) (1006779027)), time_per_zone([datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 7, 29, 996000), datetime.time(0, 1, 36, 1000), datetime.time(0, 0)] ([0, 0, 0, 0, 449996, 96001, 0])), unknown_5([invalid] ([4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295])), unknown_9([invalid] ([65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535])), hr_zones_timer_type(<HeartRateZonesTimerType.lap: 19> (19)), record_num(1), unknown_15([invalid] (65535)), hr_zones([124, 136, 148, 162, 175, 188]), hr_zones_method(<HeartRateZonesMethod.heart_rate_reserve: 2> (2)), max_heart_rate(188.0 bpm), min_heart_rate(58.0 bpm), lactate_threshhold_heart_rate(162.0 bpm), unknown_14([invalid] (255))])
Message: DataMessage(<MessageType.hr_zones_timer: 216>: [timestamp(datetime.datetime(2021, 11, 25, 12, 59, 8, tzinfo=datetime.timezone.utc) (1006779548)), time_per_zone([datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0, 59, 997000), datetime.time(0, 7, 40, 999000), datetime.time(0, 0)] ([0, 0, 0, 0, 59997, 460999, 0])), unknown_5([invalid] ([4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295])), unknown_9([invalid] ([65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535])), hr_zones_timer_type(<HeartRateZonesTimerType.lap: 19> (19)), record_num(2), unknown_15([invalid] (65535)), hr_zones([124, 136, 148, 162, 175, 188]), hr_zones_method(<HeartRateZonesMethod.heart_rate_reserve: 2> (2)), max_heart_rate(188.0 bpm), min_heart_rate(58.0 bpm), lactate_threshhold_heart_rate(162.0 bpm), unknown_14([invalid] (255))])
Message: DataMessage(<MessageType.hr_zones_timer: 216>: [timestamp(datetime.datetime(2021, 11, 25, 13, 1, 27, tzinfo=datetime.timezone.utc) (1006779687)), time_per_zone([datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0), datetime.time(0, 0, 2, 162000), datetime.time(0, 1, 59, 996000), datetime.time(0, 0)] ([0, 0, 0, 0, 2162, 119996, 0])), unknown_5([invalid] ([4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295])), unknown_9([invalid] ([65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535])), hr_zones_timer_type(<HeartRateZonesTimerType.lap: 19> (19)), record_num(3), unknown_15([invalid] (65535)), hr_zones([124, 136, 148, 162, 175, 188]), hr_zones_method(<HeartRateZonesMethod.heart_rate_reserve: 2> (2)), max_heart_rate(188.0 bpm), min_heart_rate(58.0 bpm), lactate_threshhold_heart_rate(165.0 bpm), unknown_14([invalid] (255))])
Message: DataMessage(<MessageType.hr_zones_timer: 216>: [timestamp(datetime.datetime(2021, 11, 25, 13, 1, 27, tzinfo=datetime.timezone.utc) (1006779687)), time_per_zone([datetime.time(0, 1, 9, 910000), datetime.time(0, 1, 2, 1000), datetime.time(0, 0, 39, 999000), datetime.time(0, 3, 27, 998000), datetime.time(0, 12, 21, 158000), datetime.time(0, 11, 16, 996000), datetime.time(0, 0)] ([69910, 62001, 39999, 207998, 741158, 676996, 0])), unknown_5([invalid] ([4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295, 4294967295])), unknown_9([invalid] ([65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535])), hr_zones_timer_type(<HeartRateZonesTimerType.session: 18> (18)), record_num(0), unknown_15([invalid] (65535)), hr_zones([124, 136, 148, 162, 175, 188]), hr_zones_method(<HeartRateZonesMethod.heart_rate_reserve: 2> (2)), max_heart_rate(188.0 bpm), min_heart_rate(58.0 bpm), lactate_threshhold_heart_rate(165.0 bpm), unknown_14([invalid] (255))])
```

You get data for each lap and the entire session. In particular, note this array:

`time_per_zone([datetime.time(0, 1, 9, 910000), datetime.time(0, 1, 2, 1000), datetime.time(0, 0, 39, 999000), datetime.time(0, 3, 27, 998000), datetime.time(0, 12, 21, 158000), datetime.time(0, 11, 16, 996000), datetime.time(0, 0)]`

The array is zero indexed and the 2nd through 6th entries correspond to HR zones 1-5.

I will continue to add support and capture the data in the DB.

**Comment by tcgoetz** [comment]
commit 46affdd implements addition of the following fields to the activities and laps tables:

```
    # heart rate zone data
    hr_zones_method = Column(Enum(fitfile.enum_fields.HeartRateZonesMethod))
    # heart rate threashold that the zone starts at
    hrz_1_hr = Column(Integer)
    hrz_2_hr = Column(Integer)
    hrz_3_hr = Column(Integer)
    hrz_4_hr = Column(Integer)
    hrz_5_hr = Column(Integer)
    # amount of time in that zone
    hrz_1_time = Column(Time, nullable=False, default=datetime.time.min)
    hrz_2_time = Column(Time, nullable=False, default=datetime.time.min)
    hrz_3_time = Column(Time, nullable=False, default=datetime.time.min)
    hrz_4_time = Column(Time, nullable=False, default=datetime.time.min)
    hrz_5_time = Column(Time, nullable=False, default=datetime.time.min)
```

and is currently available on the development branch. Please test and give feedback.

**Reply by smros** [reply]
I did manage to find this data and the table, fantastic!  It appears to go back 1 month or so, is that due to the program or something the time data is stored at the source?

Or, I just noticed the table (activities) is at 25 lines, matching the "latest" field in .json file?  Do I need to adjust the value in the .json file to get the heart zone values for the last year?

**Comment by smros** [comment]
That's great! When I got home I was going to start looking for those fit
files ..

If I check out the dev branch, will it access my old database? And then
would I have to update the database?

Steve

**Reply by tcgoetz** [reply]
> That's great! When I got home I was going to start looking for those fit files .. If I check out the dev branch, will it access my old database? And then would I have to update the database?

You will have to rebuild the db. If your working from a source tree, do:

`make clean reinstall rebuild_dbs`

**Comment by tcgoetz** [comment]
I added two Jupyter notebooks in commit 0c8af32. One gives a summary of all activities and the other give a summary of a single activity. They might be a good starting point for your higher level analysis.

**Comment by smros** [comment]
Awesome, thanks! Wow, you're fast, I was out skiing today and thinking
about how to write the notebooks, it'll be great to have a starting point

If I submit any of this back, do you have rules for libraries? I typically
like the pandas/bumpy, etc. tools but a lot of what I've seen so far is
more pure python. Is that what you want people to stick to?

Steve

**Reply by tcgoetz** [reply]
> If I submit any of this back, do you have rules for libraries? I typically like the pandas/bumpy, etc. tools but a lot of what I've seen so far is more pure python. Is that what you want people to stick to?

I'd like to keep the deps for the base app small, but if you do you analysis in a notebook, I have no requirements on libs for them. Just update the requirements.txt in the Jupyter directory.

**Comment by smros** [comment]
I created a clean environment and grabbed that branch.  However, I happen
to be running on a Windows machine this time (it was most convenient)
which doesn't have make (I used pip for the original install).

Is there a way to run the install with python only (i.e the setup.py file,
perhaps) or use pip with the requirements files?

If there isn't no big deal, I have Linux machines or a Mac or two I
could try as well..

-Steve

On Sun, Jan 9, 2022 at 2:44 PM Tom Goetz ***@***.***> wrote:

> That's great! When I got home I was going to start looking for those fit
> files .. If I check out the dev branch, will it access my old database? And
> then would I have to update the database?
>
> You will have to rebuild the db. If your working from a source tree, do:
>
> make clean reinstall rebuild_dbs
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tcgoetz/GarminDB/discussions/142#discussioncomment-1936798>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFYOWKFWRIP4OKLGZUMBITUVH6SJANCNFSM5KB473EA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Steve Mroszczak

E: ***@***.***
T: @smroszczak
LinkedIn: ca.linkedin.com/in/smros/

**Comment by smros** [comment]
I haven't been successful in getting a dev environment up on my Raspberry
Pi.  I tried a few things including a parallel version and a completely
clean new version, but it gets tripped up on a couple of dependencies for
the fit files and idbutils.

I modified the requirements file to just see if I could get it through the
build, but then I think it tripped as it wanted to be in the venv, which I
couldn't figure out how to activate.

What's the recommended dev environment that you use?  Ex. Ubuntu linux of a
particular version?

Thanks,

-Steve



On Sun, Jan 9, 2022 at 2:44 PM Tom Goetz ***@***.***> wrote:

> That's great! When I got home I was going to start looking for those fit
> files .. If I check out the dev branch, will it access my old database? And
> then would I have to update the database?
>
> You will have to rebuild the db. If your working from a source tree, do:
>
> make clean reinstall rebuild_dbs
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tcgoetz/GarminDB/discussions/142#discussioncomment-1936798>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFYOWKFWRIP4OKLGZUMBITUVH6SJANCNFSM5KB473EA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Steve Mroszczak

E: ***@***.***
T: @smroszczak
LinkedIn: ca.linkedin.com/in/smros/

**Reply by tcgoetz** [reply]
> What's the recommended dev environment that you use? Ex. Ubuntu linux of a particular version?

You can see my and others build setups on this page: https://github.com/tcgoetz/GarminDB/wiki/Success-Stories

**Comment by smros** [comment]
Answered my own question.  Long story, but while getting some parts for a Mac repair, I ended up with and extra SSD so setup a completely clean Ubuntu environment, and when I hit the same roadblocks on the two dependancy versions, I cloned each of those, checked out their development branches, then did a 'sudo python3 setup.py install' on each.

The setup finished and did my DB download last night, and now I am looking at the new DB.

-Steve

**Comment by tcgoetz** [comment]
You can run `make reinstall_all` at the top level to install garmindb and it's dependent packages from the tree.

**Comment by tcgoetz** [comment, marked as answer]
Heart rate zones in version 3.2.0. No current plans to add gear since its not in currently fetched Garmin Connect documents.
